src/box/lua/call.c

```diff
@@ -669,13 +669,92 @@ port_lua_dump_plain(struct port *port, uint32_t *size);
 extern struct Mem *
 port_lua_get_vdbemem(struct port *base, uint32_t *size);
 
+const struct port_c_entry *
+port_lua_get_c_entries(struct port *base)
+{
+	struct port_lua *port = (struct port_lua *)base;
+	struct lua_State *L = port->L;
+	size_t size = lua_gettop(L) - port->bottom + 1;
+	if (size == 0)
+		return NULL;
+
+	struct port_c_entry *arr =
+		xregion_alloc_array(&fiber()->gc, struct port_c_entry, size);
+
+	/* Link the list. */
+	for (size_t i = 0; i < size - 1; i++)
+		arr[i].next = &arr[i + 1];
+	arr[size - 1].next = NULL;
+
+	/* Put values. */
+	for (size_t arr_idx = 0; arr_idx < size; arr_idx++) {
+		int lua_idx = arr_idx + port->bottom;
+		struct port_c_entry *e = &arr[arr_idx];
+		switch (lua_type(L, lua_idx)) {
+		case LUA_TNIL:
+			e->type = PORT_C_ENTRY_NULL;
+			break;
+		case LUA_TBOOLEAN:
+			e->type = PORT_C_ENTRY_BOOL;
+			e->value.boolean = lua_toboolean(L, lua_idx);
+			break;
+		case LUA_TNUMBER:
+			e->type = PORT_C_ENTRY_DOUBLE;
+			e->value.number = lua_tonumber(L, lua_idx);
+			break;
+		case LUA_TSTRING: {
+			e->type = PORT_C_ENTRY_STR;
+			size_t len;
+			const char *data = lua_tolstring(L, lua_idx, &len);
+			e->value.str.data = data;
+			e->value.str.size = len;
+			break;
+		}
+		default: {
+			struct tuple *tuple = NULL;
+			tuple = luaT_istuple(L, lua_idx);
+			if (tuple != NULL) {
+				e->type = PORT_C_ENTRY_TUPLE;
+				e->value.tuple = tuple;
+				/*
+				 * Do not reference the tuple:
+				 * this entry does not own it.
+				 */
+				break;
+			}
+
+			size_t len;
+			const char *data = NULL;
+			data = luamp_get(L, lua_idx, &len);
+			if (data != NULL) {
+				e->type = PORT_C_ENTRY_MP_OBJECT;
```

Comment:
@locker 
I emphasize that I changed `PORT_C_ENTRY_MP` to `PORT_C_ENTRY_MP_OBJECT` here.